### PR TITLE
simple _cat kernel, dim 0

### PR DESCRIPTION
**Pull request by Janicewei**

### CODE REVIEW

File: hammerblade/torch/riscv.mk [hunk 1]
```diff
@@ -81,8 +81,6 @@ KERNEL_CSRCS   := $(notdir $(wildcard $(KERNEL_DIR)/*.c))
 KERNEL_CPPSRCS := $(notdir $(wildcard $(KERNEL_DIR)/*.cpp))
 KERNEL_OBJS    := $(patsubst %.c,%.o,$(KERNEL_CSRCS)) \
                   $(patsubst %.cpp,%.o,$(KERNEL_CPPSRCS))
-KERNEL_DIS     := $(patsubst %.c,%.dis,$(KERNEL_CSRCS)) \
-                  $(patsubst %.cpp,%.dis,$(KERNEL_CPPSRCS))
```
Comment (yodada) [inline hammerblade/torch/riscv.mk]: So here is the merge conflict -- I think you reverted `fb4a35c34ba6c`

You would need to revert the revert commit ;)

File: hammerblade/torch/tests/test_fallback.py [hunk 2]
```diff
@@ -41,7 +41,6 @@ def test_native_1():
     torch.hammerblade.profiler.disable()
     assert torch.allclose(hb.cpu(), cpu)
 
-@pytest.mark.xfail
 def test_native_1F():
```
Comment (yodada) [inline hammerblade/torch/tests/test_fallback.py]: Actually, can you delete this test case? It is meant to test the `fallback` mechanism and make sure it does not do crazy things when not enabled.

File: hammerblade/torch/kernel/kernel_cat.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+  bsg_cuda_print_stat_kernel_start();
+  for (size_t i = 0; i < length; i++) {
+	HBTensor<float> tensor(tensors_p[i]);
+	for (size_t j = 0; j < tensor.numel(); j++) {
+	  result(index) = tensor(j);
+	  index = index + 1;
+	}
+  }
+
```
Comment (yodada) [inline hammerblade/torch/kernel/kernel_cat.cpp:50]: Hmmm, this is a little scary -- basically you have every tile do the same work. If you are not ready to parallelize the work, you should put `if (__bsg_id == 0)`, so tile 0 will do the work while others staying idle

File: aten/src/ATen/native/hammerblade/Cat.cpp [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  scalars.push_back(create_device_scalar(dim_i32));
+
+  // XXX: this is definitely wrong
+  Tensor result = at::empty({0}, tensors[0].options());
+  bool allSkipped = true;
+  bool allContiguous = true;
+  Tensor notSkippedTensor;
+
+  // Inputs cannot alias the output tensor
+  for (size_t i = 0; i < tensors.size(); i++) {
```
Comment (yodada) [inline aten/src/ATen/native/hammerblade/Cat.cpp]: This piece of code looks strange. `result` is just allocated above `Tensor result = at::empty({0}, tensors[0].options());`, so it is not possible for it to alias with anybody.

Comment (Janicewei) [reply aten/src/ATen/native/hammerblade/Cat.cpp]: updated Cat.cpp. I implemented host code based on functions in TensorShape.cpp

File: hammerblade/torch/tests/test_cat.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    x = torch.ones(10)
+    _test_torch_cat(x)
+
+def test_cat_2():
+    x = torch.randn(3, 4)
+    _test_torch_cat(x)
+
+def test_cat_3():
+    x = torch.randn(3, 4, 5)
+    _test_torch_cat(x)
```
Comment (yodada) [inline hammerblade/torch/tests/test_cat.py:47]: More tests will make this even better ;)

For instance, I think you would want to test the case where you cat tensors with different length and empty tensors. Say, [0,1], [0,1,2,3,4,5], [] -> [0,1,0,1,2,3,4,5]

File: hammerblade/torch/kernel/kernel_cat.cpp [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  HBTensor<float> result(result_p);
+  uint32_t length = *length_p;
+  int32_t dim = *dim_p;
+  HBTensor<float> tensor(tensors_p[0]);
+  int32_t size = tensor.numel();
+  bsg_cuda_print_stat_kernel_start();
+
+  hb_tiled_for(result.numel(), [&] (size_t i) {
+    size_t j = i / size;
+    size_t index = i % size;
```
Comment (yodada) [inline hammerblade/torch/kernel/kernel_cat.cpp]: Yeah here is an assumption that isn't always true.

But as a first step, it is ok to fail on the case where input tensors have different length. However, it cannot fail silently -- you would want to check for it and throw an error

File: c10/hammerblade/CMakeLists.txt [hunk 7]
```diff
@@ -75,12 +75,12 @@ else()
   # Kernel binary
```
Comment (yodada) [inline c10/hammerblade/CMakeLists.txt]: Emmm this file shouldn't be touched. I'll help you revert the changes. Dont worry about it

File: hammerblade/torch/kernel/kernel_cat.cpp [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  bsg_cuda_print_stat_kernel_start();
+
+  hb_tiled_for(result.numel(), [&] (int32_t i) {
+    int32_t j = 0;
+    int32_t index = 0;
+    HBTensor<float> init(tensors_p[0]);
+    int32_t size = init.numel();
+    while (i >= size) {
+      index = i - size;
+      HBTensor<float> tensor(tensors_p[++j]);
```
Comment (yodada) [inline hammerblade/torch/kernel/kernel_cat.cpp]: this definitely is valid code, but personally I would avoid this. Doing `++j` inside indexing makes code hard to parse.

**Review by yodada — APPROVED** [review]
ok! Looks good! Let me fix the makefile before we merge.

File: hammerblade/torch/kernel/kernel_cat.cpp (hunk 8 above)
Comment (Janicewei) [reply hammerblade/torch/kernel/kernel_cat.cpp]: Okay! changed

**Comment by yodada** [comment]
Reverted `4f22ea6` -- now we are consistent with master branch. This PR is ready to merge once all checks are done.

@Janicewei have you tried on cosim? ideally we should give it a try before we merge

**Comment by Janicewei** [comment]
> Reverted `4f22ea6` -- now we are consistent with master branch. This PR is ready to merge once all checks are done.
> 
> @Janicewei have you tried on cosim? ideally we should give it a try before we merge

give me a few minutes

**Comment by Janicewei** [comment]
> Reverted `4f22ea6` -- now we are consistent with master branch. This PR is ready to merge once all checks are done.
> 
> @Janicewei have you tried on cosim? ideally we should give it a try before we merge

I ran `pycosim -m pytest hb-pytorch/hammerblade/torch/tests/test_cat.py -vs` and tests passed

**Comment by yodada** [comment]
> > Reverted `4f22ea6` -- now we are consistent with master branch. This PR is ready to merge once all checks are done.
> > @Janicewei have you tried on cosim? ideally we should give it a try before we merge
> 
> I ran `pycosim -m pytest hb-pytorch/hammerblade/torch/tests/test_cat.py -vs` and tests passed

perfect!

**Comment by yodada** [comment]
This PR partially implements `torch._cat` -- currently we only support `_cat` with `dim=0`
